File: src/d02/p02.rs
use std::{
    collections::HashMap,
    io::{self, Read},
};

fn main() -> io::Result<()> {
    let mut buffer: String = String::new();
    let _bytes_read = io::stdin().read_to_string(&mut buffer);

    let result = calculate_score(&buffer);

    println!("{:?}", result);

    Ok(())
}

#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
enum Move {
    Rock,
    Paper,
    Scissor,
}

use Move::*;

#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
enum Outcome {
    Win,
    Lose,
    Draw,
}

use Outcome::*;

#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug)]
enum ScoreLookupKey {
    MoveScore(Move),
    OutcomeScore(Outcome),
}

use ScoreLookupKey::*;

fn calculate_score(input: &str) -> i32 {
    let move_lookup: HashMap<char, Move> =
        HashMap::from([('A', Rock), ('B', Paper), ('C', Scissor)]);

    let desired_outcome_lookup: HashMap<char, Outcome> =
        HashMap::from([('X', Lose), ('Y', Draw), ('Z', Win)]);

    let outcome_lookup: HashMap<Move, (Move, Move, Move)> = HashMap::from([
        (Rock, (Paper, Rock, Scissor)),
        (Paper, (Scissor, Paper, Rock)),
        (Scissor, (Rock, Scissor, Paper)),
    ]);

    let score_lookup: HashMap<ScoreLookupKey, i32> = HashMap::from([
        (MoveScore(Rock), 1),
        (MoveScore(Paper), 2),
        (MoveScore(Scissor), 3),
        (OutcomeScore(Lose), 0),
        (OutcomeScore(Draw), 3),
        (OutcomeScore(Win), 6),
    ]);

    let lines = input.lines();

    let mut tot = 0;
    for line in lines {
        let opponent_char = line.chars().nth(0).unwrap();
        let desired_outcome_char = line.chars().nth(2).unwrap();

        let opponent_move = move_lookup.get(&opponent_char).unwrap();
        let desired_outcome = desired_outcome_lookup.get(&desired_outcome_char).unwrap();

        let our_move: Move = match desired_outcome {
            Lose => outcome_lookup.get(opponent_move).unwrap().2,
            Draw => outcome_lookup.get(opponent_move).unwrap().1,
            Win => outcome_lookup.get(opponent_move).unwrap().0,
        };

        let round_score = score_lookup
            .get(&OutcomeScore(desired_outcome.clone()))
            .unwrap()
            + score_lookup.get(&MoveScore(our_move.clone())).unwrap();

        tot += round_score;
    }

    return tot;
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_calculate_score() {
        let res = calculate_score("A Y\nB X\nC Z\n");
        assert_eq!(res, 12)
    }
}

File: zig/src/d02/p02.zig
const std = @import("std");

pub fn main() !void {
    // stdout is for the actual output of your application, for example if you
    // are implementing gzip, then only the compressed bytes should be sent to
    // stdout, not any debugging messages.
    const stdout_file = std.io.getStdOut().writer();
    var bw = std.io.bufferedWriter(stdout_file);
    const stdout = bw.writer();

    var inputFile = try std.fs.cwd().openFile("src/d02/input.txt", .{ .mode = .read_only });
    defer inputFile.close();

    const file_size = (try inputFile.stat()).size;
    const allocator = std.heap.page_allocator;

    var buffer = try allocator.alloc(u8, file_size);
    defer allocator.free(buffer);
    try inputFile.reader().readNoEof(buffer);

    const result = try compute_tota_score(buffer);

    try stdout.print("{d}\n", .{result});

    try bw.flush(); // don't forget to flush!
}

const MatchOutcome = enum {
    win,
    lose,
    tie,
};

const Move = enum {
    rock,
    paper,
    scissors,
};

fn getMoveScore(move: Move) i32 {
    return switch (move) {
        Move.rock => 1,
        Move.paper => 2,
        Move.scissors => 3,
    };
}

fn getOutcomeScore(outcome: MatchOutcome) i32 {
    return switch (outcome) {
        MatchOutcome.lose => 0,
        MatchOutcome.tie => 3,
        MatchOutcome.win => 6,
    };
}

const ParseMoveError = error{UnknownMove};

fn parseMove(c: u8) ParseMoveError!Move {
    return switch (c) {
        'A' => Move.rock,
        'B' => Move.paper,
        'C' => Move.scissors,
        else => ParseMoveError.UnknownMove,
    };
}

const ParseOutcomeError = error{UnknownOutcome};
fn parseOutcome(c: u8) ParseOutcomeError!MatchOutcome {
    return switch (c) {
        'X' => MatchOutcome.lose,
        'Y' => MatchOutcome.tie,
        'Z' => MatchOutcome.win,
        else => ParseOutcomeError.UnknownOutcome,
    };
}

fn computeMoveForDesiredOutcome(desiredOutcome: MatchOutcome, his: Move) Move {
    if (desiredOutcome == MatchOutcome.tie) {
        return his;
    }

    return switch (his) {
        Move.rock => if (desiredOutcome == MatchOutcome.lose) Move.scissors else Move.paper,
        Move.paper => if (desiredOutcome == MatchOutcome.lose) Move.rock else Move.scissors,
        Move.scissors => if (desiredOutcome == MatchOutcome.lose) Move.paper else Move.rock,
    };
}

fn compute_tota_score(input: []const u8) !i32 {
    var lines = std.mem.splitScalar(u8, input, '\n');
    var tot: i32 = 0;
    while (lines.next()) |line| {
        if (line.len < 3) {
            continue;
        }

        const his = try parseMove(line[0]);
        const desiredOutcome = try parseOutcome(line[2]);

        const mine = computeMoveForDesiredOutcome(desiredOutcome, his);
        tot += getMoveScore(mine) + getOutcomeScore(desiredOutcome);
    }

    return tot;
}

test "case 0" {
    const input =
        \\A Y
        \\B X
        \\C Z
        \\
    ;

    try std.testing.expectEqual(
        @as(i32, 12),
        try compute_tota_score(input),
    );
}
